sdk/core/core-rest-pipeline/test/bearerTokenAuthenticationPolicy.spec.ts

```diff
@@ -357,6 +359,527 @@ describe("BearerTokenAuthenticationPolicy", function () {
+        } catch (e) {
+          // Should not encounter an error. A request with failed status code should be returned
+          assert.fail();
+        }
+        // First getToken request will return a bad token
+        expect(getToken).toHaveBeenCalledWith(tokenScopes, {
+          abortSignal: undefined,
+          tracingOptions: undefined,
+          enableCae: true,
+        });
+        // Second getToken request will inject the correct token with the claims
+        if (testCase.expectedClaims) {
+          expect(getToken).toHaveBeenLastCalledWith(tokenScopes, {
+            enableCae: true,
+            claims: testCase.expectedClaims,
+            abortSignal: undefined,
+            tracingOptions: undefined,
+          });
+        }
+
+        if (testCase.expectedResponseCode === 200) {
+          assert.strictEqual(response.request.headers.get("Authorization"), `Bearer good-token`);
+        } else {
+          assert.strictEqual(response.request.headers.get("Authorization"), `Bearer bad-token`);
+        }
+      });
+    });
+
+    matrix([nonCaeChallengeTests] as const, async function (testCase: Challenge) {
+      // Test different scenarios when we have only 1 non-CAE challenge returned handled by a custom callback
+      it(`Non-CAE challenge test: ${testCase.testName}`, async function () {
+        const tokenExpiration = Date.now() + 1000 * 60; // One minute later.
+        const getToken = vi.fn<() => Promise<AccessToken | null>>();
+
+        // First time getToken is called will return a bad token
+        getToken.mockResolvedValueOnce({
+          token: "bad-token",
+          expiresOnTimestamp: tokenExpiration,
+        });
+        // This will return a good token after the authorization challenge is handled
+        getToken.mockResolvedValueOnce({
+          token: "good-token",
+          expiresOnTimestamp: tokenExpiration,
+        });
+        const credential: TokenCredential = {
+          getToken,
+        };
+        const scopes = ["test-scope"];
+        const request = createPipelineRequest({ url: "https://example.com" });
+
+        const challengeResponse: PipelineResponse = {
+          headers: createHttpHeaders({
+            "WWW-Authenticate": testCase.challenge,
+          }),
+          request,
+          status: 401,
+        };
+        const successResponse: PipelineResponse = {
+          headers: createHttpHeaders(),
+          request,
+          status: 200,
+        };
+
+        let isCallbackCalled = false;
+        async function authorizeRequestOnChallenge(
+          options: AuthorizeRequestOnChallengeOptions,
+        ): Promise<boolean> {
+          isCallbackCalled = true;
+          assert.equal(testCase.challenge, options.response.headers.get("WWW-Authenticate"));
+          // Should set the good token here in the second get access token
+          const token = await options.getAccessToken(scopes, {
+            claims: testCase.challenge,
```

Comment:
What's the motivation for this? It looks strange because today a client should set `claims` only when handling a CAE challenge

Reply:
Ah that's my misunderstanding when setting up the challenge callback. Updated!